### PR TITLE
add masked fill op

### PR DESCRIPTION
**Pull request by clackhan**
添加masked fill op
reference https://pytorch.org/docs/stable/tensors.html?highlight=masked_fill#torch.Tensor.masked_fill

### CODE REVIEW

File: oneflow/user/ops/masked_fill_op.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    .Input("x")
+    .Input("mask")
+    .Attr("value", UserOpAttrType::kAtFloat)
+    .Output("y")
+    .SetTensorDescInferFn([](user_op::InferContext* ctx) -> Maybe<void> {
+      const user_op::TensorDesc* x = ctx->TensorDesc4ArgNameAndIndex("x", 0);
+      const user_op::TensorDesc* mask = ctx->TensorDesc4ArgNameAndIndex("mask", 0);
+      user_op::TensorDesc* y = ctx->TensorDesc4ArgNameAndIndex("y", 0);
+      size_t output_num_axes = std::max(x->shape().NumAxes(), mask->shape().NumAxes());
+      if (IsScalarTensor(mask)) {
```
Comment (liujuncheng) [inline oneflow/user/ops/masked_fill_op.cpp:37]: 这里是要求mask可以broadcast成x还是x,y之间可以相互broadcast

File: oneflow/user/ops/masked_fill_op.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    })
+    .SetBatchAxisInferFn(user_op::BatchAxisInferFnUtil::NaiveInferBatchAxis)
+    .SetInputArgModifyFn([](user_op::GetInputArgModifier GetInputArgModifierFn,
+                            const user_op::UserOpConfWrapper&) {
+      user_op::InputArgModifier* cond_arg_modifier = GetInputArgModifierFn("mask", 0);
+      cond_arg_modifier->set_requires_grad(false);
+    })
+    .SetGetSbpFn(user_op::GetSbpFnUtil::SplitForEachAxis);
+
+REGISTER_USER_OP("masked_fill_grad")
```
Comment (liujuncheng) [inline oneflow/user/ops/masked_fill_op.cpp:67]: masked_fill的后向是不是也是masked_fill

File: oneflow/user/kernels/masked_fill_kernel.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  void Compute(user_op::KernelComputeContext* ctx) const override {
+    const user_op::Tensor* x = ctx->Tensor4ArgNameAndIndex("x", 0);
+    const user_op::Tensor* mask = ctx->Tensor4ArgNameAndIndex("mask", 0);
+    user_op::Tensor* y = ctx->Tensor4ArgNameAndIndex("y", 0);
+    user_op::Tensor* tmp_mask_and_y = ctx->Tensor4ArgNameAndIndex("tmp_buffer", 0);
+    T value = static_cast<T>(ctx->Attr<float>("value"));
+    const auto mask_shape = mask->shape();
+    const auto y_shape = y->shape();
+    size_t num_axes = y_shape.NumAxes();
+    KernelUtil<DeviceType::kCPU, T>::AddByScalar(ctx->device_ctx(), mask_shape.elem_cnt(),
```
Comment (liujuncheng) [inline oneflow/user/kernels/masked_fill_kernel.cpp:52]: 如果在python端用
```python
values = flow.constant_like(like=x, value=value)
return flow.where(condition=mask, x=values, y=x)
```
是不是也可以实现masked_fill

File: oneflow/user/kernels/masked_fill_kernel.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                                                 mask->dptr<T>(), static_cast<T>(-1),
```
Comment (liujuncheng) [inline oneflow/user/kernels/masked_fill_kernel.cpp:53]: mask的类型可能和T不一样吧

File: oneflow/user/kernels/masked_fill_kernel.cpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                       & (user_op::HobDataType("dy", 0) == GetDataType<dtype>::value)) \
+      .SetInferTmpSizeFn([](user_op::InferContext* ctx) {                              \
+        const Shape* mask_shape = ctx->Shape4ArgNameAndIndex("mask", 0);               \
+        const Shape* y_shape = ctx->Shape4ArgNameAndIndex("dy", 0);                    \
+        return GetCudaAlignedSize(mask_shape->elem_cnt() * sizeof(dtype))              \
+               + GetCudaAlignedSize(y_shape->elem_cnt() * sizeof(dtype))               \
+               + GetCudaAlignedSize(y_shape->elem_cnt() * sizeof(dtype));              \
+      });
+
+REGISTER_CPU_MASKED_FILL_GRAD_KERNEL(float)
```
Comment (liujuncheng) [inline oneflow/user/kernels/masked_fill_kernel.cpp:148]: GPU要支持half

File: oneflow/user/ops/masked_fill_op.cpp (hunk 1 above)
Comment (clackhan) [reply oneflow/user/ops/masked_fill_op.cpp:37]: 要求x和mask可以相互broadcast，y的形状是x和mask推导得到。

File: oneflow/user/kernels/masked_fill_kernel.cpp (hunk 4 above)
Comment (clackhan) [reply oneflow/user/kernels/masked_fill_kernel.cpp:53]: mask是bool只，在系统里用int8处理，这里还要优化。

File: oneflow/user/ops/masked_fill_op.cpp (hunk 2 above)
Comment (clackhan) [reply oneflow/user/ops/masked_fill_op.cpp:67]: 还是不一样的：

x: [2,4], mask: [[1], [0], [0]], value: 10
前向得到：
y=[[10, 10], [2, 4], [2, 4]]
y_diff = [[2, 4], [6, 8], [3, 2]], 后向得到：
x_diff = [9,10]，会对前向没有填充的部分做reduce_sum

File: oneflow/user/kernels/masked_fill_kernel.cpp (hunk 5 above)
Comment (clackhan) [reply oneflow/user/kernels/masked_fill_kernel.cpp:148]: 目前添加的两个op都没支持half，打算把基本类型写好后，在支持half

File: oneflow/python/ops/array_ops.py [hunk 6]
```diff
@@ -879,3 +879,16 @@ def broadcast_like(
         .Build()
     )
     return op.InferAndTryRun().SoleOutputBlob()
+
+
+@oneflow_export("masked_fill")
```
Comment (liujuncheng) [inline oneflow/python/ops/array_ops.py:884]: 补充一下文档？

File: oneflow/python/test/ops/test_masked_fill.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    np_out, np_x_diff = _masked_fill_np_fw_bw(x, mask, out_diff, np_type, value)
+
+    test_case.assertTrue(np.allclose(np_out, test_global_storage.Get("out")))
+    test_case.assertTrue(np.allclose(np_x_diff, test_global_storage.Get("x_diff")))
+
+
+def test_masked_fill_fw_bw(test_case):
+    arg_dict = OrderedDict()
+    arg_dict["type_name"] = ["float32", "double", "int8", "int32", "int64"]
```
Comment (liujuncheng) [inline oneflow/python/test/ops/test_masked_fill.py:117]: 添加一下fp16的支持和测试用例，应该要要给constant_like和where添加fp16支持

File: oneflow/core/kernel/broadcast_to_compatible_with_kernel.cpp [hunk 8]
```diff
@@ -58,4 +58,8 @@ void BroadcastToCompatibleWithKernel<device_type, T>::ForwardDataContent(
 OF_PP_SEQ_PRODUCT_FOR_EACH_TUPLE(REGISTTER_BROADCAST_TO_COMPATIBLE_WITH_KERNEL, DEVICE_TYPE_SEQ,
                                  ARITHMETIC_DATA_TYPE_SEQ)
 
+#if defined(WITH_CUDA)
+REGISTTER_BROADCAST_TO_COMPATIBLE_WITH_KERNEL(DeviceType::kGPU, (float16, DataType::kFloat16))
```
Comment (liujuncheng) [inline oneflow/core/kernel/broadcast_to_compatible_with_kernel.cpp:62]:
```suggestion
REGISTTER_BROADCAST_TO_COMPATIBLE_WITH_KERNEL(DeviceType::kGPU, FLOAT16_DATA_TYPE_SEQ)
```

File: oneflow/core/kernel/constant_like_kernel_half.cu [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+#include "oneflow/core/common/data_type.h"
+#include "oneflow/core/kernel/kernel.h"
+#include "oneflow/core/kernel/new_kernel_util.h"
+#include "oneflow/core/kernel/util/cuda_half_util.h"
+
+namespace oneflow {
+
+namespace {
+
+__global__ void NaiveHalfFillGpu(const int64_t elem_cnt, const float16 x, float16* y) {
```
Comment (liujuncheng) [inline oneflow/core/kernel/constant_like_kernel_half.cu:25]: `NewKernelUtil<device_type>::Fill`是支持fp16的，是不是只需要处理好operand到float16的转换，而不是新写kernel

File: oneflow/core/kernel/constant_like_kernel_half.cu [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  }
+};
+
+#define REGISTER_HALF_CONSTANT_LIKE_KERNEL                                                 \
+  REGISTER_KERNEL_WITH_DEVICE_AND_DTYPE(OperatorConf::kConstantLikeConf, DeviceType::kGPU, \
+                                        float16, HalfConstantLikeKernel)
+
+REGISTER_HALF_CONSTANT_LIKE_KERNEL
+
+}  // namespace oneflow
```
Comment (liujuncheng) [inline oneflow/core/kernel/constant_like_kernel_half.cu:69]: 注意这里的空行

File: oneflow/user/kernels/scalar_by_tensor_add_kernel_half.cu [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+limitations under the License.
+*/
+#include "oneflow/core/framework/framework.h"
+#include "oneflow/core/kernel/util/cuda_half_util.h"
+
+namespace oneflow {
+
+namespace {
+
+__global__ void HalfAddByScalarPtrGpu(const int64_t n, const half* x, const half* y, half* z) {
```
Comment (liujuncheng) [inline oneflow/user/kernels/scalar_by_tensor_add_kernel_half.cu:23]: 这里应该是给XxxByScalarPtr添加float16类型吧

File: oneflow/user/kernels/where_kernel_half.cu [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+*/
+#include "oneflow/core/framework/framework.h"
+#include "oneflow/core/kernel/util/cuda_half_util.h"
+
+namespace oneflow {
+
+namespace {
+
+template<typename CondT>
+__global__ void NaiveHalfWhere(const int64_t elem_cnt, const CondT* cond, const half* lhs,
```
Comment (liujuncheng) [inline oneflow/user/kernels/where_kernel_half.cu:24]: 同理，这里直接给where添加float16就可以了吧

File: oneflow/user/kernels/scalar_by_tensor_add_kernel_half.cu [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  const half y_value = y[0];
+  CUDA_1D_KERNEL_LOOP(i, n) { z[i] = x[i] + y_value; }
```
Comment (liujuncheng) [inline oneflow/user/kernels/scalar_by_tensor_add_kernel_half.cu:25]: float16/half类型不能直接用`+`，要用__hadd

File: oneflow/core/kernel/broadcast_to_compatible_with_kernel.cpp (hunk 8 above)
Comment (clackhan) [reply oneflow/core/kernel/broadcast_to_compatible_with_kernel.cpp]: 使用宏FLOAT16_DATA_TYPE_SEQ的话会导致REGISTTER_BROADCAST_TO_COMPATIBLE_WITH_KERNEL将FLOAT16_DATA_TYPE_SEQ本身作为参数，从而发生宏展开错误的情况，故仍使用(float16, DataType::kFloat16)

Comment (liujuncheng) [reply oneflow/core/kernel/broadcast_to_compatible_with_kernel.cpp]: > 使用宏FLOAT16_DATA_TYPE_SEQ的话会导致REGISTTER_BROADCAST_TO_COMPATIBLE_WITH_KERNEL将FLOAT16_DATA_TYPE_SEQ本身作为参数，从而发生宏展开错误的情况，故仍使用(float16, DataType::kFloat16)

好

**Review by liujuncheng — APPROVED** [review]
